File: iron_bar.swift
#!/usr/bin/env swift
// http://programmingpraxis.com/2015/10/06/the-iron-bar/

func ironBar(n: Int...) -> (actualMedian: Int, ironBarMedian: Int) {
  let offset = n.count % 2 == 1 ? 0 : 1
  let actualMedian = n.sort { $0 < $1 }[n.count/2-offset]
  var ironBarMedian = n[0]
  for i in 1..<n.count {
    ironBarMedian += n[i] > ironBarMedian ? 1 : n[i] < ironBarMedian ? -1 : 0
  }
  return (actualMedian, ironBarMedian)
}

print(ironBar(29, 26, 31, 30, 30, 28, 28, 31, 24, 32))

File: iron_bar.rb
#!/usr/bin/env ruby
# http://programmingpraxis.com/2015/10/06/the-iron-bar/

def iron_bar(*n)
  actual_median = n.count.odd? ?
    "n.sort[n.count/2]" : "(n.sort[(n.count/2)-1] + n.sort[(n.count/2)]) / 2.0"
  iron_bar_median = n[0]
  (1...n.count).each do |i| 
    iron_bar_median += n[i] > iron_bar_median ? 1 : n[i] < iron_bar_median ? -1 : 0
  end
  { "actual_median" => eval(actual_median), "iron_bar_median" => iron_bar_median }
end

p iron_bar(29, 26, 31, 30, 30, 28, 28, 31, 24, 32)
